File: polyglot/python/license_parser.py
"""
polyglot/python/license_parser.py

A production-grade SPDX and file-based license parser for licenselens.
Detects, normalizes, and validates software licenses from various sources.

Usage:
    >>> import json
    >>> result = parse_license("MIT License\nCopyright (c) 2024...")
    >>> print(json.dumps(result, indent=2))
"""

import re
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Optional, Dict, Any, List, Tuple
from pathlib import Path


class LicenseStatus(Enum):
    """Detection confidence levels."""
    EXACT = "exact"           # Full match with all key terms
    HIGH = "high"             # 90%+ confidence
    MEDIUM = "medium"         # 75-89% confidence  
    LOW = "low"               # 60-74% confidence
    UNKNOWN = "unknown"       # Minimal or no indicators


@dataclass(frozen=True)
class LicenseMatch:
    """Single license detection result."""
    name: str
    status: LicenseStatus
    spdx_id: Optional[str]
    source_lines: List[int]  # Line numbers in original text
    confidence_score: float
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "status": self.status.value,
            "spdx_id": self.spdx_id,
            "source_lines": self.source_lines,
            "confidence_score": round(self.confidence_score, 2),
        }


@dataclass
class ParseResult:
    """Complete parsing result for one license source."""
    raw_text: str
    detected_licenses: List[LicenseMatch] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "raw_length": len(self.raw_text),
            "detected_licenses": [l.to_dict() for l in self.detected_licenses],
            "warnings": self.warnings,
            "summary": f"{len(self.detected_licenses)} license(s) detected",
        }


class LicenseParser:
    """
    Core license detection engine.
    
    Supports: SPDX identifiers, common file headers (MIT, Apache 2.0, BSD, GPL),
    and URL-based lookups via public APIs.
    """
    
    # Known licenses with high-confidence patterns
    KNOWN_LICENSES = {
        "MIT": {
            "spdx_id": "MIT",
            "patterns": [
                r"MIT License",
                r"(?:\b|_)MIT\b.*?License",
                r"\(c\s*\d{4}\s*,?\s*The\s+MIT\s+Licence",
            ],
        },
        "Apache-2.0": {
            "spdx_id": "Apache-2.0",
            "patterns": [
                r"Apache License.*?Version 2\.0",
                r"(?:\b|_)Apache\s+License.*?2\.0",
                r"\(c\s*\d{4}\s*,?\s*The\s+Apache\s+Software\s+Foundation",
            ],
        },
        "BSD-3-Clause": {
            "spdx_id": "BSD-3-Clause",
            "patterns": [
                r"BSD.*?License.*?3-clause",
                r"(?:\b|_)BSD.*?3\s*Clause",
            ],
        },
        "GPL-3.0-only": {
            "spdx_id": "GPL-3.0-only",
            "patterns": [
                r"GNU General Public License.*?Version 3\.0",
                r"(?:\b|_)GPL.*?v3",
            ],
        },
    }
    
    # Fallback: common license name patterns for fuzzy matching
    FUZZY_PATTERNS = {
        "MIT": [r"mit\s*license", r"simple\s*permission"],
        "Apache-2.0": [r"apache.*?version 2", r"asf\s+foundation"],
        "BSD-3-Clause": [r"bsd.*?3-clause", r"freebsd\s+base"],
        "GPL-3.0-only": [r"gpl.*?v3", r"gnu\s+general\s+public\s+license.*?3\.0"],
    }
    
    def __init__(self, fuzzy_threshold: float = 0.75):
        """
        Initialize parser with configuration.
        
        Args:
            fuzzy_threshold: Minimum score for fuzzy matches (0-1)
        """
        self.fuzzy_threshold = fuzzy_threshold
    
    def parse(self, text: str, source_hint: Optional[str] = None) -> ParseResult:
        """
        Parse license from raw text.
        
        Args:
            text: Raw license content or file header
            source_hint: Optional hint like "file_header" or "spdx_string"
            
        Returns:
            Complete parse result with detected licenses and warnings
        """
        result = ParseResult(raw_text=text)
        
        # Normalize input for better matching
        normalized = self._normalize(text)
        
        # Step 1: Try exact pattern matches against known licenses
        exact_matches = []
        for name, config in self.KNOWN_LICENSES.items():
            for pattern in config["patterns"]:
                match = re.search(pattern, text, re.IGNORECASE | re.MULTILINE)
                if match:
                    # Calculate confidence based on how much of the expected content matched
                    score = self._calculate_exact_score(match.group(0), name, normalized)
                    exact_matches.append(LicenseMatch(
                        name=name,
                        status=LicenseStatus.EXACT if score > 0.9 else LicenseStatus.HIGH,
                        spdx_id=config["spdx_id"],
                        source_lines=[],  # Would need line tracking for files
                        confidence_score=score,
                    ))
                    break
        
        # Step 2: Fuzzy matching if no exact matches found
        if not exact_matches and normalized.strip():
            fuzzy_results = self._fuzzy_match(normalized)
            exact_matches.extend(fuzzy_results)
        
        # Step 3: Check for SPDX identifier directly in text
        spdx_id = self._extract_spdx_identifier(text)
        if spdx_id:
            config = next(
                (c for c in self.KNOWN_LICENSES.values() 
                 if c["spdx_id"] == spdx_id), None
            ) or {"name": "Unknown", "spdx_id": spdx_id}
            
            # Only add if not already detected with higher confidence
            existing = [m for m in exact_matches if m.name.lower() == config["name"].lower()]
            if not existing:
                exact_matches.insert(0, LicenseMatch(
                    name=config["name"],
                    status=LicenseStatus.MEDIUM,
                    spdx_id=spdx_id,
                    source_lines=[],
                    confidence_score=0.65,
                ))
        
        # Deduplicate and sort by confidence
        exact_matches = self._deduplicate(exact_matches)
        exact_matches.sort(key=lambda x: -x.confidence_score)
        
        result.detected_licenses = exact_matches
        
        return result
    
    def _normalize(self, text: str) -> str:
        """Normalize text for better pattern matching."""
        # Remove excessive whitespace while preserving structure
        lines = [line.strip() for line in text.split('\n')]
        normalized_lines = []
        
        for i, line in enumerate(lines):
            if not line or line.isspace():
                continue  # Skip empty lines
            
            # Normalize common variations
            line_lower = line.lower()
            
            # Remove URL fragments that might interfere
            url_pattern = r'https?://[^\s<>"{}|\\^`\[\]]+'
            clean_line = re.sub(url_pattern, 'URL', line)
            
            normalized_lines.append(clean_line)
        
        return '\n'.join(normalized_lines).strip()
    
    def _calculate_exact_score(self, matched_text: str, license_name: str, 
                              normalized: str) -> float:
        """Calculate confidence score for an exact match."""
        base = 0.85
        
        # Bonus for common copyright year format
        if re.search(r'\(c\s*\d{4}\s*,?\s*The', matched_text):
            base += 0.1
            
        # Check if license name appears in the match
        if license_name.lower() in matched_text.lower():
            base = min(base + 0.05, 0.98)
        
        return round(min(base, 1.0), 2)
    
    def _fuzzy_match(self, normalized: str) -> List[LicenseMatch]:
        """Perform fuzzy matching against known licenses."""
        results = []
        
        for name, patterns in self.FUZZY_PATTERNS.items():
            total_score = 0.0
            matches_found = 0
            
            for pattern in patterns:
                if re.search(pattern, normalized, re.IGNORECASE):
                    # Score based on how many key terms matched
                    score = len(re.findall(r'\w+', pattern)) / max(3, len(re.findall(r'\w+', pattern)))
                    total_score += score
                    matches_found += 1
            
            if matches_found > 0:
                avg_score = total_score / matches_found
                
                # Determine status based on score and number of matches
                if avg_score >= 0.85 or matches_found == len(patterns):
                    status = LicenseStatus.HIGH
                elif avg_score >= 0.75:
                    status = LicenseStatus.MEDIUM
                else:
                    status = LicenseStatus.LOW
                
                results.append(LicenseMatch(
                    name=name,
                    status=status,
                    spdx_id=self.KNOWN_LICENSES.get(name, {}).get("spdx_id"),
                    source_lines=[],
                    confidence_score=round(avg_score, 2),
                ))
        
        return results
    
    def _extract_spdx_identifier(self, text: str) -> Optional[str]:
        """Extract SPDX identifier if present in the text."""
        # Look for explicit SPDX ID declarations
        spdx_pattern = r'(?:SPDX-|LicenseRef-)?[A-Za-z0-9._\-/]+(?:\s*#.*?)?'
        
        matches = re.findall(spdx_pattern, text, re.IGNORECASE)
        
        if matches:
            # Return the most likely SPDX ID (first non-prefixed one)
            for match in matches:
                clean_match = re.sub(r'SPDX-|LicenseRef-', '', match, flags=re.IGNORECASE)
                if clean_match and clean_match not in ['SPDX', 'LICENSEREF']:
                    return clean_match
        
        return None
    
    def _deduplicate(self, matches: List[LicenseMatch]) -> List[LicenseMatch]:
        """Remove duplicate license detections."""
        seen = set()
        unique = []
        
        for match in matches:
            key = (match.name.lower(), match.spdx_id or "")
            
            if key not in seen:
                seen.add(key)
                unique.append(match)
        
        return unique


def parse_license(text: str, fuzzy_threshold: float = 0.75) -> ParseResult:
    """
    Convenience function for parsing a single license text.
    
    Args:
        text: Raw license content
        fuzzy_threshold: Minimum score for fuzzy matches
        
    Returns:
        Complete parse result
    """
    parser = LicenseParser(fuzzy_threshold=fuzzy_threshold)
    return parser.parse(text)


def parse_file(path: Union[str, Path], fuzzy_threshold: float = 0.75) -> ParseResult:
    """
    Parse license from a file path.
    
    Args:
        path: File path or Path object
        fuzzy_threshold: Minimum score for fuzzy matches
        
    Returns:
        Complete parse result
    """
    parser = LicenseParser(fuzzy_threshold=fuzzy_threshold)
    with open(path, 'r', encoding='utf-8') as f:
        return parser.parse(f.read(), source_hint="file")


def main():
    """Demo and self-test for the license parser."""
    
    # Test cases covering common scenarios
    
    test_cases = [
        ("MIT License", "MIT License\nCopyright (c) 2024 Example Corp.\nPermission is hereby granted..."),
        ("Apache-2.0", "Apache License, Version 2.0\nLicensed under the Apache License, Version 2.0..."),
        ("BSD-3-Clause", "Redistribution and use in source and binary forms, permitted so long as the following conditions are met:\n(1) Redistributions must retain the above copyright notice..."),
        ("GPL-3.0-only", "GNU GENERAL PUBLIC LICENSE\nVersion 3, 29 June 2007\nCopyright (C) 2007 Free Software Foundation, Inc."),
        ("Unknown/Custom", "Some Custom License\nCopyright 2024 My Company"),
        ("Empty", ""),
        ("Minimal MIT", "MIT License"),
    ]
    
    print("=" * 60)
    print("License Parser Self-Test")
    print("=" * 60)
    
    for name, text in test_cases:
        print(f"\n--- Test: {name} ---")
        result = parse_license(text)
        
        if result.detected_licenses:
            for lic in result.detected_licenses[:3]:  # Show top 3 matches
                print(f"  ✓ {lic.name}: {lic.status.value} ({lic.confidence_score})")
        else:
            print("  ✗ No licenses detected")
        
        if result.warnings:
            for warning in result.warnings[:2]:
                print(f"    ⚠ Warning: {warning}")


if __name__ == "__main__":
    main()

File: polyglot/rust/license_parser.rs
// polyglot/rust/license_parser.rs

use std::collections::{HashMap, HashSet};
use std::fmt;
use std::fs;
use std::path::PathBuf;

/// Represents a parsed license with metadata and source information.
#[derive(Debug, Clone)]
pub struct LicenseInfo {
    pub id: String,
    pub name: String,
    pub spdx_id: Option<String>,
    pub osi_approved: bool,
    pub url: Option<String>,
    pub text: Option<String>,
    pub confidence: f32,
}

/// Result of parsing a license source.
#[derive(Debug, Clone)]
pub struct ParseResult {
    pub info: LicenseInfo,
    pub source_type: SourceType,
    pub raw_input: String,
    pub warnings: Vec<String>,
}

/// Types of sources that can be parsed.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SourceType {
    File(PathBuf),
    Stdin,
    SpdxIdentifier(String),
    RawString,
}

impl fmt::Display for SourceType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SourceType::File(p) => write!(f, "file({})", p.display()),
            SourceType::Stdin => write!(f, "stdin"),
            SourceType::SpdxIdentifier(s) => write!(f, "spdx_identifier(\"{}\")", s),
            SourceType::RawString => write!(f, "raw_string"),
        }
    }
}

/// Known licenses database.
#[derive(Debug)]
pub struct LicenseDatabase {
    /// Exact name matches (case-insensitive).
    exact: HashMap<String, String>, // name -> spdx_id
    /// Partial/fuzzy patterns.
    partial: Vec<(String, String)>, // pattern -> spdx_id
    /// OSI approved list.
    osi_approved: HashSet<String>,
}

impl Default for LicenseDatabase {
    fn default() -> Self {
        let mut exact = HashMap::new();
        let mut partial = Vec::new();
        
        // Exact matches (common ones)
        exact.insert("mit".to_ascii_lowercase(), "MIT".to_string());
        exact.insert("apache 2.0".to_ascii_lowercase(), "Apache-2.0".to_string());
        exact.insert("apache-2.0".to_ascii_lowercase(), "Apache-2.0".to_string());
        exact.insert("apache2".to_ascii_lowercase(), "Apache-2.0".to_string());
        exact.insert("bsd 3-clause".to_ascii_lowercase(), "BSD-3-Clause".to_string());
        exact.insert("bsd-3-clause".to_ascii_lowercase(), "BSD-3-Clause".to_string());
        exact.insert("isc".to_ascii_lowercase(), "ISC".to_string());
        exact.insert("zlib".to_ascii_lowercase(), "Zlib".to_string());
        exact.insert("unlicense".to_ascii_lowercase(), "Unlicense".to_string());
        exact.insert("wtfpl".to_ascii_lowercase(), "WTFPL".to_string());
        exact.insert("cc0-1.0".to_ascii_lowercase(), "CC0-1.0".to_string());
        
        // Partial matches with patterns
        partial.push(("gnu general public license v3".to_ascii_lowercase(), "GPL-3.0-only".to_string()));
        partial.push(("gplv3".to_ascii_lowercase(), "GPL-3.0-only".to_string()));
        partial.push(("affero general public license".to_ascii_lowercase(), "AGPL-3.0-only".to_string()));
        
        // OSI approved licenses (subset for demo)
        let mut osi = HashSet::new();
        osi.insert("MIT".to_string());
        osi.insert("Apache-2.0".to_string());
        osi.insert("BSD-3-Clause".to_string());
        osi.insert("ISC".to_string());
        osi.insert("Zlib".to_string());
        osi.insert("Unlicense".to_string());
        osi.insert("WTFPL".to_string());
        osi.insert("CC0-1.0".to_string());
        
        LicenseDatabase { exact, partial, osi_approved: osi }
    }
}

impl LicenseDatabase {
    pub fn new() -> Self {
        Default::default()
    }
    
    /// Check if a license name is OSI approved.
    pub fn is_osi_approved(&self, spdx_id: &str) -> bool {
        self.osi_approved.contains(spdx_id.to_ascii_uppercase().as_str())
    }

    /// Get the canonical SPDX ID for a license name.
    pub fn resolve_spdx_id(&self, input: &str) -> Option<String> {
        let normalized = input.trim().to_ascii_lowercase();
        
        // First try exact match
        if let Some(spdx) = self.exact.get(&normalized) {
            return Some(spdx.clone());
        }

        // Try partial pattern matching
        for (pattern, spdx) in &self.partial {
            if normalized.contains(pattern.as_str()) || pattern.contains(&normalized) {
                return Some(spdx.clone());
            }
        }

        None
    }
}

/// Parse a license from various sources.
pub fn parse_license(
    source: SourceType,
    raw_input: &str,
    db: &LicenseDatabase,
) -> Result<ParseResult, ParseError> {
    let info = match &source {
        SourceType::SpdxIdentifier(spdx_id) => {
            LicenseInfo {
                id: spdx_id.clone(),
                name: format!("SPDX ID: {}", spdx_id),
                spdx_id: Some(spdx_id.clone()),
                osi_approved: db.is_osi_approved(spdx_id),
                url: None,
                text: None,
                confidence: 1.0,
            }
        }
        
        SourceType::Stdin | SourceType::RawString => {
            parse_text(raw_input, db)
        }
        
        SourceType::File(path) => {
            let content = fs::read_to_string(path).map_err(|e| ParseError::Io(e))?;
            parse_text(&content, db)
        }
    };

    Ok(ParseResult {
        info,
        source_type: source,
        raw_input: raw_input.to_string(),
        warnings: Vec::new(),
    })
}

/// Parse license text content.
fn parse_text(content: &str, db: &LicenseDatabase) -> LicenseInfo {
    let normalized = content.trim().to_ascii_lowercase();
    
    // Check for SPDX identifier at start of file
    if let Some(spdx_id) = extract_spdx_header(&normalized) {
        return LicenseInfo {
            id: spdx_id.clone(),
            name: format!("SPDX Header: {}", spdx_id),
            spdx_id: Some(spdx_id),
            osi_approved: db.is_osi_approved(&spdx_id),
            url: None,
            text: Some(content.to_string()),
            confidence: 0.95,
        };
    }

    // Check for common license headers/patterns
    if let Some(spdx) = db.resolve_spdx_id(&normalized) {
        return LicenseInfo {
            id: spdx.clone(),
            name: format!("Detected: {}", spdx),
            spdx_id: Some(spdx),
            osi_approved: db.is_osi_approved(&spdx),
            url: None,
            text: Some(content.to_string()),
            confidence: 0.85,
        };
    }

    // Check for common license name patterns (case-insensitive)
    let detected = detect_license_name(&normalized);
    
    LicenseInfo {
        id: format!("detected-{}", detected),
        name: detected.clone(),
        spdx_id: None,
        osi_approved: false,
        url: None,
        text: Some(content.to_string()),
        confidence: 0.6,
    }
}

/// Extract SPDX identifier from file header.
fn extract_spdx_header(normalized: &str) -> Option<String> {
    // Look for "SPDX-License-Identifier:" pattern
    if let Some(pos) = normalized.find("spdx-license-identifier:") {
        let after = &normalized[pos + 24..];
        let id = after.split_whitespace().next()?.trim();
        
        // Validate it looks like an SPDX ID
        if id.starts_with('A') || id.starts_with('B') || 
           id.starts_with('C') || id.starts_with('E') ||
           id.starts_with('G') || id.starts_with('I') ||
           id.starts_with('L') || id.starts_with('M') ||
           id.starts_with('P') || id.starts_with("CC-") {
            return Some(id.to_string());
        }
    }

    // Look for "SPDX-FileCopyrightText:" followed by license reference
    if let Some(pos) = normalized.find("spdx-filecopyrighttext:") {
        let after = &normalized[pos + 25..];
        if let Some(eq_pos) = after.find('=') {
            let id = after[eq_pos + 1..].split_whitespace().next()?.trim();
            return Some(id.to_string());
        }
    }

    None
}

/// Detect license name from text content.
fn detect_license_name(normalized: &str) -> String {
    // Common license name patterns (partial matches)
    let patterns = [
        ("mit", "MIT License"),
        ("apache 2.0", "Apache License 2.0"),
        ("bsd 3-clause", "BSD 3-Clause License"),
        ("bsd 2-clause", "BSD 2-Clause License"),
        ("isc", "ISC License"),
        ("zlib", "Zlib License"),
        ("unlicense", "The Unlicense"),
        ("wtfpl", "WTFPL"),
        ("cc0-1.0", "CC0 Public Domain"),
        ("gnu gpl v3", "GNU GPL v3.0"),
        ("gplv3", "GNU GPL v3.0"),
        ("gnu lpl v2", "GNU LGPL v2.1"),
        ("gplv2", "GNU GPL v2.0"),
        ("gnu gpl v2", "GNU GPL v2.0"),
    ];

    for (pattern, name) in &patterns {
        if normalized.contains(*pattern) || pattern.contains(&normalized) {
            return name.to_string();
        }
    }

    // Fallback: look for license-like text patterns
    if normalized.contains("permission is hereby granted") {
        return "Permissive License".to_string();
    }

    if normalized.contains("redistribution and use in source and binary forms") {
        return "Standard Open Source License".to_string();
    }

    // Generic fallback
    let words: Vec<&str> = normalized.split_whitespace().collect();
    for word in &words[..10] {
        if !word.is_empty() && word.len() > 3 {
            return format!("Unknown (starts with '{}')", word);
        }
    }

    "Unknown".to_string()
}

/// Errors that can occur during parsing.
#[derive(Debug)]
pub enum ParseError {
    Io(std::io::Error),
    EmptyInput,
    InvalidSpdx(String),
    UnknownSource(String),
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ParseError::Io(e) => write!(f, "IO error: {}", e),
            ParseError::EmptyInput => write!(f, "No input provided"),
            ParseError::InvalidSpdx(s) => write!(f, "Invalid SPDX identifier: {}", s),
            ParseError::UnknownSource(s) => write!(f, "Unknown source type: {}", s),
        }
    }
}

impl std::error::Error for ParseError {}

/// Builder for creating a license parser with custom database.
pub struct LicenseParserBuilder {
    db: LicenseDatabase,
}

impl Default for LicenseParserBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl LicenseParserBuilder {
    pub fn new() -> Self {
        Self {
            db: LicenseDatabase::default(),
        }
    }

    /// Build the parser with current database.
    pub fn build(self) -> LicenseParser {
        LicenseParser { db: self.db }
    }

    /// Add a custom license mapping.
    pub fn add_mapping(&mut self, name: &str, spdx_id: &str) {
        let normalized = name.to_ascii_lowercase();
        self.db.exact.insert(normalized, spdx_id.to_string());
    }

    /// Load additional license data from a file (e.g., JSON database).
    pub fn load_from_json(&mut self, path: impl AsRef<std::path::Path>) -> Result<(), ParseError> {
        let content = fs::read_to_string(path.as_ref())?;
        
        // Simple JSON parsing for demo purposes
        if content.contains("\"spdx_id\"") && content.contains("\"name\"") {
            // Extract SPDX IDs and names from the file
            let spdx_ids: Vec<&str> = content
                .split('"')
                .filter(|s| s.starts_with("SPDX-"))
                .collect();
            
            for id in &spdx_ids[..10] {
                if !id.is_empty() && id.len() > 5 {
                    let spdx = format!("{}-{}", id[4..].trim(), "demo");
                    self.db.exact.insert(id.to_ascii_lowercase(), spdx);
                }
            }
        }

        Ok(())
    }
}

/// The main license parser.
pub struct LicenseParser {
    db: LicenseDatabase,
}

impl Default for LicenseParser {
    fn default() -> Self {
        Self::new()
    }
}

impl LicenseParser {
    pub fn new() -> Self {
        Self {
            db: LicenseDatabase::default(),
        }
    }

    /// Parse a license from the given source.
    pub fn parse(&self, source: SourceType, raw_input: &str) -> Result<ParseResult, ParseError> {
        self.db.resolve_spdx_id(raw_input).map(|_| {
            let info = LicenseInfo {
                id: "demo".to_string(),
                name: format!("Demo Parser"),
                spdx_id: Some("MIT".to_string()),
                osi_approved: true,
                url: None,
                text: Some(raw_input.to_string()),
                confidence: 0.9,
            };

            ParseResult {
                info,
                source_type: source,
                raw_input: raw_input.to_string(),
                warnings: Vec::new(),
            }
        })
    }

    /// Check if a license is OSI approved.
    pub fn is_osi_approved(&self, spdx_id: &str) -> bool {
        self.db.is_osi_approved(spdx_id)
    }

    /// Get the canonical SPDX ID for input.
    pub fn resolve_spdx_id(&self, input: &str) -> Option<String> {
        self.db.resolve_spdx_id(input)
    }

    /// Parse multiple licenses from a directory (SBOM generation).
    pub fn parse_directory<P: AsRef<std::path::Path>>(
        &self,
        dir_path: P,
    ) -> Result<Vec<ParseResult>, ParseError> {
        let mut results = Vec::new();

        for entry in fs::read_dir(dir_path)? {
            let entry = entry?;
            let path = entry.path();

            if path.is_file() {
                let extension = path.extension().and_then(|e| e.to_str());
                
                // Only process text-based license files
                match extension {
                    Some("txt") | Some("md") | Some("rst") | Some("license") => {
                        let content = fs::read_to_string(&path)?;
                        
                        if !content.is_empty() {
                            let source = SourceType::File(path.clone());
                            
                            // Skip very small files (might be binary or metadata)
                            if content.len() > 100 {
                                match parse_license(source, &content, &self.db) {
                                    Ok(result) => results.push(result),
                                    Err(e) => {
                                        eprintln!("Error parsing {}: {}", path.display(), e);
                                    }
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }
        }

        Ok(results)
    }

    /// Generate a simple SBOM from parsed licenses.
    pub fn generate_sbom(&self, results: &[ParseResult]) -> String {
        let mut sbom = String::from(r#"SPDX-JSON-Format-Version: 2.0
SPDXID: SPDXRef-DOCUMENT
DocumentNamespace